src/submodule.c

```diff
@@ -149,6 +149,45 @@ static int find_by_path(const git_config_entry *entry, void *payload)
 	return 0;
 }
 
+/*
+ * Checks to see if the submodule shares its name with a file or directory that
+ * already exists on the index. If so, the submodule cannot be added.
+ */
+static int is_path_occupied(bool *occupied, git_repository *repo, const char *path)
+{
+	int error = 0;
+	git_index *index;
+	git_buf dir = GIT_BUF_INIT;
+	*occupied = false;
+
+	if ((error = git_repository_index__weakptr(&index, repo)) < 0)
+		goto out;
+
+	if ((error = git_index_find(NULL, index, path)) == 0) {
+		giterr_set(GITERR_SUBMODULE,
+			"File '%s' already exists in the index", path);
+		*occupied = true;
+		goto out;
+	}
+
+	if ((error = git_buf_sets(&dir, path)) < 0)
+		goto out;
+
+	if ((error = git_path_to_dir(&dir)) < 0)
+		goto out;
+
+	if ((error = git_index_find_prefix(NULL, index, dir.ptr)) == 0) {
```

Comment:
You're ignoring any potential error returned by `git_index_find_prefix` other than `GIT_ENOTFOUND`. Granted, right now it will not return any other error, but you never know how it will change in the future. Maybe something like this:

```
if ((error = git_index_find_prefix(NULL, index, dir.ptr)) < 0 && error != GIT_ENOTFOUND)
    goto out;

if (!error) {
    giterr_set(GITERR_SUBMODULE, ...);
    *occupied = true;
    goto out;
}

error = 0;
```